Docs: Add docs for `vue3-vite` framework

kylegach opened the pull request:
## What I did

See title

![Screenshot of docs page](https://github.com/storybookjs/storybook/assets/486540/793d47f7-d1f7-4104-9402-b981d5dac5e9)

## Checklist for Contributors

### Testing

#### Manual testing

1. Follow the steps in the [contributing instructions](https://storybook.js.org/docs/react/contribute/new-snippets#preview-your-work) for this branch, `framework-doc-vue3-vite`

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [x] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

kylegach committed:
Add docs for `vue3-vite` framework

larsrickert commented on docs/get-started/vue-vite.md:
```diff
@@ -0,0 +1,268 @@
+
+## Using `vue-component-meta`
+
+[`vue-component-meta`](https://github.com/vuejs/language-tools/tree/master/packages/component-meta) is a tool that extracts metadata from Vue components. Storybook can use it to generate the [controls](../essentials/controls.md) for your stories and documentation. It's a more full-featured alternative to `vue-docgen-api` and is recommended for most projects.
```
```suggestion
[`vue-component-meta`](https://github.com/vuejs/language-tools/tree/master/packages/component-meta) is a tool that extracts metadata from Vue components which is maintained by the official Vue team. Storybook can use it to generate the [controls](../essentials/controls.md) for your stories and documentation. It's a more full-featured alternative to `vue-docgen-api` and is recommended for most projects.

It requires Storybook >= 8. `vue-component-meta` is currently opt-in but will become the default in future versions of Storybook.
```

larsrickert commented on docs/get-started/vue-vite.md:
```diff
@@ -0,0 +1,268 @@
+
+### Support for multiple component types
+
+`vue-component-meta` supports from all types of Vue components, including functional, composition / options API components, and from `.vue`, `.ts`, `.tsx`, `.js`, and `.jsx` files.
```
```suggestion
`vue-component-meta` supports all types of Vue components (including SFC, functional, composition / options API components) from `.vue`, `.ts`, `.tsx`, `.js`, and `.jsx` files.
```

larsrickert commented on docs/get-started/vue-vite.md:
````diff
@@ -0,0 +1,268 @@
+```html
+<!-- YourComponent.vue -->
+<script setup lang="ts">
+  interface MyComponentProps {
+    /**
+     * The name of the user
+     */
+    name: string;
+    /**
+     * The category of the component
+     * @required
+     * @default 'Uncategorized'
+     */
+    category: string;
+  }
````
```suggestion
interface MyComponentProps {
  /**
   * The name of the user
   */
  name: string;
  /**
   * The category of the component
   *
   * @since 8.0.0
   */
  category?: string;
}

withDefaults(defineProps<MyComponentProps>(), {
  category: "Uncategorized",
});
```

Required will be automatically detected so for`category?: string` it will be false and true for `category: string`.
Also the default will probably be inferred by `withDefaults` for the most projects, right?

larsrickert commented on docs/get-started/vue-vite.md:
```diff
@@ -0,0 +1,268 @@
+
+<Callout variant="info">
+
+Due to technical limitations of Volar / vue language features, JSDoc descriptions for emitted events can not be extracted.
```
This is technically true but I have implemented a workaround for this so event descriptions work even with `vue-component-meta`.

Basically the event descriptions will be extracted with `vue-docgen-api` even if using `vue-component-meta`.

See
https://github.com/storybookjs/storybook/blob/a010d45c8b0b323631ade0f167b84f8a97313818/code/frameworks/vue3-vite/src/plugins/vue-component-meta.ts#L209

larsrickert replied:
Maybe you can even add descriptions in the code snippet below and update the screenshot?

larsrickert commented on docs/get-started/vue-vite.md:
```diff
@@ -0,0 +1,268 @@
+
+The definition above will generate the following controls:
+
+![Controls generated from slots](./vue-component-meta-slot-types-controls.png)
```
```suggestion
![Controls generated from slots](./vue-component-meta-slot-types-controls.png)

If using `defineSlots`, the controls will even include the description:
```ts
defineSlots<{
  /** Example description no-bind. */
  noBind(props: {}): any;
  /** Example description default. */
  default(props: { num: number }): any;
  /** Example description named. */
  named(props: { str: string }): any;
  /** Example description vbind. */
  vbind(props: { num: number; str: string }): any;
}>();
```
```

larsrickert commented on docs/get-started/vue-vite.md:
````diff
@@ -0,0 +1,268 @@
+export default config;
+```
+
+#### `docgen`
````
This is only supported since version 8, right? At the top of the file it says `>= 7`

kylegach replied:
Good call. The framework itself is supported from 7+, but this API is only in 8+. I've clarified that.

larsrickert commented on docs/get-started/vue-vite.md:
````diff
@@ -0,0 +1,268 @@
+// Rest of the file...
+```
+
+## Using `vue-component-meta`
````
Should we also be transparent about current Volar limitations? Something like:

Due to a [vue-component-meta limitation](https://github.com/vuejs/language-tools/issues/3896), `tsconfig.json` files that include `references` are currently not supported. This can lead to missing resolved types/controls if you are using import aliases (`compilerOptions.paths` in your tsconfig).

Until the feature request is implemented in `vue-component-meta`, you can switch to relative imports for types used for props, emits, slots or exposed:
```ts
import type { MyProps } from "@/types"; // can not be resolved
import type { MyProps } from "../types"; // can be resolved
```

larsrickert commented on docs/get-started/vue-vite.md:
````diff
@@ -0,0 +1,268 @@
+```html
+<!-- YourComponent.vue -->
+<script setup lang="ts">
+  type MyFooEvent = 'foo';
+
+  interface MyEvents {
+    (event: MyFooEvent, data?: { foo: string }): void;
+    (event: 'bar', value: { year: number; title?: any }): void;
+    (e: 'baz'): void;
+  }
+
+  const emit = defineEmits<MyEvents>();
````
```suggestion
type MyFooEvent = 'foo';

interface MyEvents {
  (event: MyFooEvent, data?: { foo: string }): void;
  (event: 'bar', value: { year: number; title?: any }): void;
  (e: 'baz'): void;
}

const emit = defineEmits<MyEvents>();
```

Personal preference to remove the indentation :)

kylegach replied:
Prettier indents them. 🙃 The only way to leave them un-indented would be to have prettier ignore the block. That could cause us to publish a malformed snippet, so I don't think it's worth that trade-off.

larsrickert commented on docs/get-started/vue-vite.md:
````diff
@@ -0,0 +1,268 @@
+```html
+<!-- YourComponent.vue -->
+<script setup lang="ts">
+  import { ref } from 'vue';
+
+  const label = ref('Button');
+  const count = ref(100);
+
+  defineExpose({
+    /**
+     * a label string
+     */
+    label,
+    /**
+     * a count number
+     */
+    count,
+  });
````
```suggestion
import { ref } from 'vue';

const label = ref('Button');
const count = ref(100);

defineExpose({
  /**
   * a label string
   */
  label,
  /**
   * a count number
   */
  count,
});
```

Personal preference to remove the indentation :)

kylegach commented:
@chakAs3 @larsrickert — Hello! This my attempt at officially documented what you added in #/22285. You might notice that a lot of is copy/pasted from your excellent PR description. 🙂 But I'd appreciate your thoughts, anyway!

kylegach committed:
Apply suggestions from code review

Co-authored-by: Lars Rickert <dev@lars-rickert.de>

kylegach committed:
Address comments

- Update URL & sidebar nav
- Improve snippets and update screenshots
- Clarify that `docgen` API is only available in SB 8+
- Remove no descriptions limitation from events
- Add limitations section

kylegach committed:
Add builder options details

kylegach committed:
Address TKs

kylegach committed:
Make content conditional on required renderer

kylegach committed:
Replace README

kylegach committed:
Add troubleshooting

kylegach committed:
Rename snippet
